### PR TITLE
fix(typecheck): improve error message when tsc outputs help text

### LINKED ISSUE #8981: Running `vitest` with `--typecheck` option throws error that prints out `tsc --help` if `tsconfig.json` is not available
### Describe the bug

I'm not 100% sure if this classifies as a bug, but it certainly classifies as horrible DX. If we assume that a "working" error message should tell me what went wrong, then it's certainly a bug.

If I run `vitest --typecheck`, I get the following error:

```sh
> vitest run --typecheck

Testing types with tsc and vue-tsc is an experimental feature.
Breaking changes might not follow SemVer, please pin Vitest's version when using it.

 RUN  v4.0.8 /home/foo

 ✓  TS  __tests__/exports/bar.test-d.ts (1 test)
   ✓ Usage in ESM context (1)
     ✓ Module exposes type `Bar`
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Typecheck Error ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error: Version 5.9.3
tsc: The TypeScript Compiler - Version 5.9.3

COMMON COMMANDS

  tsc
  Compiles the current project (tsconfig.json in the working directory.)

  tsc app.ts util.ts
  Ignoring tsconfig.json, compiles the specified files with default compiler options.

  tsc -b
  Build a composite project in the working directory.

  tsc --init
  Creates a tsconfig.json with the recommended settings in the working directory.

  /// THE WHOLE PRINTOUT OF `tsc --help` CONTINUES ///

You can learn about all of the compiler options at https://aka.ms/tsc


⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯


 Test Files  1 passed (1)
      Tests  1 passed (1)
Type Errors  no errors
     Errors  1 error
   Start at  17:05:43
   Duration  118ms (transform 0ms, setup 0ms, collect 0ms, tests 0ms, environment 0ms, prepare 0ms, typecheck 81ms)  
```

So the test itself succeeds, but some error occurs and the error message does not give any hint about the actual issue.

After a lengthy investigation, it turned out that this is caused by the fact that I do not have a `tsconfig.json` file, but several different tsconfig files with different names.

Proposal: if the `tsconfig.json` (or the config file that is configured via `tests.typecheck.tsconfig` is not found,  vitest should print an error that says that that file was not found and how it can be configured.

### Reproduction

Take any vitest setup with working typecheck configured.

Remove/rename the `tsconfig.json`

### System Info

```shell
Node: 22.12.0
npm: 10.9.0

"typescript": "^5.9.3",
"vitest": "4.0.8"
```

### Used Package Manager

npm

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vitest.dev/guide/).
- [x] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by Ujjwaljain16**
When tsconfig.json is missing and tsc cannot find any configuration file (including in parent directories), it outputs its help text instead of type checking errors. This was being captured and displayed as a confusing 'Typecheck Error' with the entire help text dump (100+ lines of tsc usage instructions).

Changes:
1. Added help text detection in prepareResults() method
   - Checks for 'The TypeScript Compiler - Version' header
   - Checks for 'COMMON COMMANDS' section
   - Both are present in tsc help output

2. Throws clear, actionable error message with:
   - Explanation of what went wrong
   - The configured tsconfig path (or default 'tsconfig.json')
   - Three possible solutions: a. Ensure tsconfig exists in project root b. Verify custom tsconfig path in vitest config c. Check tsconfig is valid JSON

3. Added stderr capture in spawn() method
   - Previously only stdout was captured
   - Configuration errors from tsc may appear on stderr
   - Ensures all tsc output is examined for help text

Testing:
- All existing tests pass (7/7 in test/typescript)
- Manually verified tsc outputs help text when tsconfig missing
- Automated test not included due to tsc's parent directory search behavior making it impractical to reliably trigger this scenario

Note: This is different from TS5058 error which occurs when a specific tsconfig path is provided but doesn't exist. The help text scenario happens when tsc can't locate any config file at all.

### Description

Resolves #8981

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.
- [ ] Please check [Allow edits by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to make review process faster. Note that this option is not available for repositories that are owned by Github organizations.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: test/snapshots/test/fixtures/summary/basic.test.ts [hunk 1]
```diff
@@ -1,15 +1,15 @@
 import { expect, test } from 'vitest'
```
Comment (sheremet-va) [inline test/snapshots/test/fixtures/summary/basic.test.ts:1]: What are these changes?

**Review by sheremet-va — CHANGES_REQUESTED** [review]
Please, add a test that confirms the behaviour is now correct.

**Comment by Ujjwaljain16** [comment]
sorry for the confusion!
Those snapshot changes were accidentally included when i staged all files i have updated the PR to remove them  the commit now only contains the typecheck fix in [typechecker.ts]

**Comment by Ujjwaljain16** [comment]
I've added unit tests in
 ```test/typescript/test/typechecker.test.ts``` 
that directly verify the help text detection behavior:

Detects help text with version marker - Confirms the error thrown includes the configured tsconfig path and actionable solutions
Detects help text with COMMON COMMANDS marker - Ensures detection works with partial help output
Normal tsc output passes through - Verifies regular type errors aren't affected by the new detection logic
All tests pass (3 new + 7 existing = 10/10). I opted for unit tests rather than integration tests since it's difficult to reliably trigger the help text scenario in an end-to-end test (tsc searches parent directories for config files)

**Comment by Ujjwaljain16** [comment]
Both CI failures (throws an error if typechecker process exists in typescript tests and boolean flag 100 should not crash CLI in config tests) appear to be pre-existing flaky tests unrelated to this PR. My changes only touch:

```typechecker.ts (help text detection)```
```[typechecker.test.ts] (new unit tests - all passing)```


The failing tests are in completely different areas (test runner error handling and coverage reporting).

File: test/snapshots/test/fixtures/summary/basic.test.ts (hunk 1 above)
Comment (Ujjwaljain16) [reply test/snapshots/test/fixtures/summary/basic.test.ts:1]: sorry for the confusion!
Those snapshot changes were accidentally included when i staged all files i have updated the PR to remove them the commit now only contains the typecheck fix in [typechecker.ts]

**Comment by Ujjwaljain16** [comment]
@sheremet-va  waiting for yr review

**Comment by Ujjwaljain16** [comment]
The failing test "boolean flag 100 should not crash CLI" is unrelated to the typecheck changes in this PR. This PR only modifies typecheck/typechecker.ts for error handling, while the failure is in coverage threshold reporting. The test is expecting "ERROR: Coverage for functions" but it's not appearing in stderr.

All typecheck-related tests pass successfully, including the previously failing Windows CI test

File: test/typescript/test/typechecker.test.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
+import type { TestProject } from '../../../packages/vitest/src/node/project'
```
Comment (sheremet-va) [inline test/typescript/test/typechecker.test.ts]: This test doesn't actually test the functionality, it mocks everything. Please, add a test following the policy from AGENTS.md:

https://github.com/vitest-dev/vitest/blob/60642b30482cd95fb56abe4e0250ad8beaa5467a/AGENTS.md#L41

Create a fixture or use `runInlineTests` to run tests.

Comment (Ujjwaljain16) [reply test/typescript/test/typechecker.test.ts]: The “help text” scenario cannot be integration-tested because `tsc` always
searches parent directories for tsconfig.json files. Even with runInlineTests,
tsc still inherits a real working directory and climbs to the Vitest monorepo
root, where tsconfig files exist.

Because of this, tsc will never emit its help text in CI or in Vitest’s test
environment, and automated tests cannot reliably trigger this condition.
@sheremet-va

Comment (sheremet-va) [reply test/typescript/test/typechecker.test.ts]: Can't you just generate an invalid tsconfig file then?

Comment (Ujjwaljain16) [reply test/typescript/test/typechecker.test.ts]: @sheremet-va Thank you for the suggestion! I attempted the invalid tsconfig approach but ran into a technical constraint I'd like your input on.

## Problem: tsc Help Text Trigger Conditions

After testing, I found that `tsc` only outputs help text under these conditions:
- Invoked with zero arguments: `tsc`
- Invoked with unrecognized flags: `tsc --unknownFlag`
- Invoked with `--help`

However, Vitest's typechecker always invokes tsc with valid arguments:
```typescript
// From typechecker.ts spawn()
const args = [
  '--noEmit',
  '--pretty', 'false',
  '--incremental',
  '--tsBuildInfoFile', '...',
  '-p', resolve(root, typecheck.tsconfig)
]
const child = x(typecheck.checker, args, {...})
```

## Invalid `tsconfig.json` Attempts

| Approach | `runInlineTests` Config | `tsc` Output | Result |
|--------|-------------------------|--------------|--------|
| **Syntactically invalid JSON** | `'tsconfig.json': '{ invalid }'` | N/A — Vite/esbuild fails first | ❌ Build error before `tsc` runs |
| **Empty object** | `'tsconfig.json': '{}'` | `error TS18003: No inputs were found` | ❌ Error output, not help text |
| **Missing `compilerOptions`** | `'tsconfig.json': '{"files":[]}'` | `error TS18003: No inputs were found` | ❌ Error output, not help text |

### Root Cause

When `tsc` is invoked with `-p <file>` and the config file is **structurally valid JSON** (even if semantically incorrect), it enters normal compiler mode. In this mode, `tsc` emits **specific diagnostic errors** and never falls back to printing help text.

Help text is only produced when `tsc` itself is invoked with **invalid or missing CLI arguments**, which does not occur in Vitest’s invocation path.


Current Solution: Test Executable Stub
I implemented a temporary executable that outputs tsc help text:
```
// Creates cross-platform executable
const fakeTscPath = path.join(os.tmpdir(), `fake-tsc-${Date.now()}`)
fs.writeFileSync(fakeTscPath, '#!/usr/bin/env node\nconsole.log("Version 5.3.3");\n...')
fs.chmodSync(fakeTscPath, '755')

// Configure Vitest to use it
const config = `export default defineConfig({
  test: { typecheck: { checker: '${fakeTscPath}' } }
})`

```

Is there a specific tsconfig.json structure or tsc invocation pattern I'm missing that would trigger help text output when called with -p <file>?
Alternatively, is the test executable approach acceptable for this edge case, or would you prefer:


A different testing strategy you have in mind
Modifying how Vitest invokes tsc for this specific test scenario

File: test/typescript/test/typecheck-error.test.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    // SOLUTION: Use a test executable that mimics tsc help output
+    // This is NOT a mock (no jest.mock or similar), but a real executable script
+    // that Vitest spawns and executes, validating the error handling logic works.
+
+    const fs = await import('node:fs')
+    const path = await import('node:path')
+    const os = await import('node:os')
+
+    // Create a temporary directory for our fake tsc
+    const tmpDir = path.join(os.tmpdir(), `vitest-test-${Date.now()}`)
```
Comment (sheremet-va) [inline test/typescript/test/typecheck-error.test.ts:20]: I like this test, but we have `createFile` utility in `test/test-utils/index` that will remove the file when the test is finished. Currently, the file will only be removed if `runInlineTests` succeds

File: test/typescript/test/typecheck-error.test.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    // 2. Vitest always calls tsc with proper arguments (--noEmit, --pretty, etc.)
+    // 3. Invalid tsconfig causes ERROR output, not help text
+    //
+    // SOLUTION: Use a test executable that mimics tsc help output
+    // This is NOT a mock (no jest.mock or similar), but a real executable script
+    // that Vitest spawns and executes, validating the error handling logic works.
+
+    const fs = await import('node:fs')
+    const path = await import('node:path')
+    const os = await import('node:os')
```
Comment (sheremet-va) [inline test/typescript/test/typecheck-error.test.ts]: I don't see any reason to import these dynamically

**Comment by Ujjwaljain16** [comment]
@sheremet-va Thank you for the feedback! I've addressed both points:

## Changes Made

**1.  Static Imports**
Replaced dynamic imports with static imports at the top of the file:
```typescript
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import { createFile, runInlineTests } from '../../test-utils'
```
**2. Using createFile Utility**
Replaced manual file creation and cleanup with the createFile() utility:
```
createFile(fakeTscPath, scriptContent)
```
***Test Results***
```
✓ test/typecheck-error.test.ts (1 test) 417ms
  ✓ Typechecker Error Handling (1)
    ✓ throws helpful error when tsc outputs help text (missing config)

Test Files  1 passed (1)
Tests       1 passed (1)
````

File: packages/vitest/src/typecheck/typechecker.ts [hunk 5]
```diff
@@ -123,6 +123,23 @@ export class Typechecker {
...
+      const tsconfigPath = typecheck.tsconfig || 'tsconfig.json'
+      const msg = `TypeScript compiler returned help text instead of type checking results.\n`
+        + `This usually means the tsconfig file was not found.\n\n`
+        + `Possible solutions:\n`
+        + `  1. Ensure '${tsconfigPath}' exists in your project root\n`
+        + `  2. If using a custom tsconfig, verify the path in your Vitest config:\n`
+        + `     test: { typecheck: { tsconfig: 'path/to/tsconfig.json' } }\n`
+        + `  3. Check that the tsconfig file is valid JSON`
+
+      console.error(msg)
```
Comment (sheremet-va) [inline packages/vitest/src/typecheck/typechecker.ts]:
```suggestion
```

File: test/typescript/test/typecheck-error.test.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    // 2. Vitest always calls tsc with proper arguments (--noEmit, --pretty, etc.)
+    // 3. Invalid tsconfig causes ERROR output, not help text
+    //
+    // SOLUTION: Use a test executable that mimics tsc help output
+    // This is NOT a mock (no jest.mock or similar), but a real executable script
+    // that Vitest spawns and executes, validating the error handling logic works.
+
+    // Create a temporary directory for our fake tsc
+    const tmpDir = path.join(os.tmpdir(), `vitest-test-${Date.now()}`)
+    fs.mkdirSync(tmpDir, { recursive: true })
```
Comment (sheremet-va) [inline test/typescript/test/typecheck-error.test.ts]: `createFile` already does this

```suggestion
```

Comment (Ujjwaljain16) [reply test/typescript/test/typecheck-error.test.ts]: thanks for review changes made

**Comment by sheremet-va** [comment]
Thank you!
